### PR TITLE
Improve documentation of Hash

### LINKED ISSUE #7593: Is Hash ordered based on when the entry was inserted?
Hi, everybody.

  I just want to make sure whether `Hash#keys` or `Hash#each` returns entries in order based on when they were inserted. I found it's mentioned in [`Hash#key_index`](https://crystal-lang.org/api/0.27.2/Hash.html#key_index(key)) in document,  and it looks like it's ordered viewing the code of `key_index`. But nothing mentioned in any other place about this, and the document of [`Set` ](https://crystal-lang.org/api/0.27.2/Set.html) say `Set` based on `Hash` is unordered.

### PR DESCRIPTION
**Pull request by straight-shoota**
Closes #7593

### CODE REVIEW

File: src/hash.cr [hunk 1]
```diff
@@ -397,6 +425,8 @@ class Hash(K, V)
   #   key_and_value # => {"foo", "bar"}
   # end
   # ```
+  #
+  # The enumeration follows the order the keys were inserted.
```
Comment (Sija) [inline src/hash.cr]: I'd add it as `NOTE`, ditto all below.

File: src/hash.cr [hunk 2]
```diff
@@ -1,8 +1,36 @@
...
+# ```
+#
+# Implementation is based on an open hash table.
+# Two objects refer to the same hash key when their hash value (`Object#hash`)
+# is identical and both objects are equal to each other (`Object#==`).
+#
+# Enumeration follows the order that the corresponding keys were inserted.
+#
+# NOTE: When using mutable data types as keys, changing the value of a key after
+# it was inserted into the `Hash` may lead to undefined behaviour.
```
Comment (asterite) [inline src/hash.cr]: Optional: maybe mention that in those cases you can call `rehash`?

File: src/hash.cr (hunk 1 above)
Comment (straight-shoota) [reply src/hash.cr]: I don't think these labels should be overused.

File: src/hash.cr [hunk 3]
```diff
@@ -1,8 +1,37 @@
...
 
-# A `Hash` represents a mapping of keys to values.
+# A `Hash` represents a collection of key-value mappings, similar to a dictionary.
 #
-# See the [official docs](http://crystal-lang.org/docs/syntax_and_semantics/literals/hash.html) for the basics.
+# Main operations are storing a key-value mapping (`#[]=`) and
+# querying the value associated to a key (`#[]`). Key-value mappings can also be
+# deleted (`#delete`).
+# Keys are unique within a hash. When adding a key-value mapping with a key that
+# is already in use, the old value will be forgotten.
```
Comment (r00ster91) [inline src/hash.cr]:
```suggestion
# is already in use, the old value will be overwritten.
```

Comment (asterite) [reply src/hash.cr]: I'm not sure, "overwritten" sounds like the actual value will be modified, but that's not the case. I think "forgotten" is fine.

Comment (Sija) [reply src/hash.cr]: Maybe _replaced_ would do?

**Pull request merged**